File: pyNA/src/noise_src_py/split_subbands.py
import numpy as np
from typing import Dict, Any


def split_subbands(settings: Dict[str, Any], msap_in: np.ndarray) -> np.ndarray:
    """
    Compute subfrequency bands for a given 1/3rd octave frequency spectrum.

    :param settings: pyna settings
    :type settings: Dict[str, Any]
    :param msap_in: mean-square acoustic pressure of the source (re. rho_0,^2c_0^2) [-]
    :type msap_in: np.ndarray [settings['n_frequency_bands']']]

    :return: mean-square acoustic pressure of the source, split into sub-frequency bands (re. rho_0,^2c_0^2) [-]
    :rtype: msap_sb [settings['n_frequency_bands']']*settings['n_frequency_subbands']']]

    """

    # Integer [-]
    m = (settings['n_frequency_subbands'] - 1) / 2

    # Initialize counter
    cntr = -1

    msap_sb = np.zeros(settings['n_frequency_bands'] * settings['n_frequency_subbands'])

    for k in np.arange(settings['n_frequency_bands']):

        # Extract msap_k
        msap_in_k = msap_in[k]

        # Normalize the mean-square acoustic pressure: divide the vector by the average value
        # Computational fix for equations for u / v
        if sum(msap_in) == 0:
            msap_proc = msap_in
        else:
            msap_proc = msap_in / (np.sum(msap_in) / msap_in.shape[0])

        if msap_proc[k] == 0:
                msap_sb[k * settings['n_frequency_subbands']:(k + 1) * settings['n_frequency_subbands']] = (np.sum(msap_in) ** 0) * np.zeros([settings['n_frequency_subbands']])
        else:
            # Compute slope of spectrum
            # Source: Zorumski report 1982 part 1. Chapter 5.1 Equation 8-9
            if 0 < k < settings['n_frequency_bands'] - 1:
                u = msap_proc[k] / msap_proc[k - 1]
                v = msap_proc[k + 1] / msap_proc[k]
            elif k == 0:
                u = msap_proc[1] / msap_proc[0]
                v = msap_proc[1] / msap_proc[0]
            elif k == settings['n_frequency_bands'] - 1:
                u = msap_proc[k] / msap_proc[k - 1]
                v = msap_proc[k] / msap_proc[k - 1]

            # Compute constant A
            # Source: Zorumski report 1982 part 1. Chapter 5.1 Equation 12 + Berton ground-effects paper
            A = 1
            for h in np.arange(1, m + 1):
                A = A + u ** ((h - m - 1) / settings['n_frequency_subbands']) + v ** ((h) / settings['n_frequency_subbands'])

            # Compute MSAP in sub-bands
            # Source: Zorumski report 1982 part 1. Chapter 5.1 Equation 10 + Berton ground-effects paper
            for h in np.arange(settings['n_frequency_subbands']):
                # Update counter
                cntr = cntr + 1

                # Compute subband msap
                if 0 <= h <= m - 1:
                    msap_sb[cntr] = (msap_in_k / A) * u ** ((h - m) / settings['n_frequency_subbands'])
                elif h == m:
                    msap_sb[cntr] = (msap_in_k / A)
                elif m + 1 <= h <= settings['n_frequency_subbands'] - 1:
                    msap_sb[cntr] = (msap_in_k / A) * v ** ((h - m) / settings['n_frequency_subbands'])

    return msap_sb

File: pyNA/src/noise_src_jl/split_subbands.jl
function split_subbands!(spl_sb, spl, settings)
    
    # x = spl
    # y = spl_sb
    
    # Integer for subband calculation
    m = Int64((settings["n_frequency_subbands"] - 1) / 2)
 
    # Calculate slope of spectrum [1, N_f]
    # Source: Zorumski report 1982 part 1. Chapter 5.1 Equation 8-9
    for i in 1:1:settings["n_frequency_bands"]
                
        if i == 1
            u = spl[2]/spl[1]
            v = spl[2]/spl[1]
        elseif i == settings["n_frequency_bands"]
            u = spl[end]/spl[end - 1]
            v = spl[end]/spl[end - 1]
        else
            u = spl[i]/spl[i-1] 
            v = spl[i+1]/spl[i]
        end
        
        A = 1
        for h in 1:m
            A += u^((h-m-1)/settings["n_frequency_subbands"]) + v^(h/settings["n_frequency_subbands"])
        end
        
        spl_sb[(i-1)*settings["n_frequency_subbands"] .+ (1:2)] = spl[i]/A .* u.^ (((1:2) .- m .-1) / settings["n_frequency_subbands"])
        spl_sb[(i-1)*settings["n_frequency_subbands"] .+ 3]     = spl[i]/A
        spl_sb[(i-1)*settings["n_frequency_subbands"] .+ (4:5)] = spl[i]/A .* v.^ (((4:5) .- m .-1) / settings["n_frequency_subbands"])
        
    end
    
end

split_subbands_fwd! = (y, x)->split_subbands!(y, x, settings)